File: ants/bots/java/HunterBot.java
import java.util.*;

public class HunterBot implements Runnable {
	private static Ants ants = new Ants();
	public static void main(String[] args) {
		Ants.run(new HunterBot(), ants);
	}

	public void run() {
		Set<Tile> destinations = new HashSet<Tile>();
		Set<Tile> targets = new HashSet<Tile>();
		targets.addAll(ants.food());
		targets.addAll(ants.enemyAnts());
		for (Tile location : ants.myAnts()) {
			boolean issued = false;
			Tile closestTarget = null;
			int closestDistance = 999999;
			for (Tile target : targets) {
				int distance = ants.distance(location, target);
				if (distance < closestDistance) {
					closestDistance = distance;
					closestTarget = target;
				}
			}
			if (closestTarget != null) {
				List<Aim> directions = ants.directions(location, closestTarget);
				Collections.shuffle(directions);
				for (Aim direction : directions) {
					Tile destination = ants.tile(location, direction);
					if (ants.ilk(destination).isUnoccupied() && !destinations.contains(destination)) {
						ants.issueOrder(location, direction);
						destinations.add(destination);
						issued = true;
						break;
					}
				}
			}
			if (!issued) {
				destinations.add(location);
			}
		}
	}
}

File: ants/bots/scala/src/main/scala/HunterBot.scala
object HunterBot extends Application {
  new AntsGame().run(new HunterBot)
}

class HunterBot extends Bot {

  def ordersFrom(game: Game): Set[Order] = {

    // Your logic goes here.
    // for example ...

    val targets = (game.board.food ++ game.board.enemyAnts).values.toList

    val antAndDirections = game.board.myAnts.values.map{ant =>
      val distance: (Positionable) => Double = (p) => game.distanceFrom(ant.tile).to(p.tile)
      val maybeTarget: Option[Positionable] = targets.sortWith(distance(_) < distance(_)).headOption
      val maybeDirection: Option[CardinalPoint] = maybeTarget.flatMap(target => game.directionsFrom(ant.tile).to(target.tile).headOption)
      (ant, maybeDirection)
    }.filter(_._2.isDefined).map{case(ant,maybeDirection) => (ant, maybeDirection.get)}

    val validOrders = antAndDirections.flatMap{case(ant,direction) =>
      val targetTile = game.tile(direction).of(ant.tile)
      game.board.elements.get(targetTile) match {
        case None | Some(Corpse(_)) => Some(Order(ant.tile, direction))
        case _ => None
      }
    }

    validOrders.toSet
  }

}